### PR TITLE
fix: copy logs when changing BNs to strings

### PR DESCRIPTION
**Pull request by mrice32**
**Motivation**

The logger is currently modifying objects that are passed to it when it contains a BigNumber or BN.

**Summary**

This change updates the logger to detect when a nested value has changed and then copying all containing objects recursively.

**Details**

This may be unnecessary for some PRs. Catch-all for detailed explanations about the implementation decisions and implications of the change.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [x]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**

N/A

### CODE REVIEW

**Review by nicholaspai — APPROVED** [review]
What kind of bad consequences were arising without this fix?

**Comment by mrice32** [comment]
> What kind of bad consequences were arising without this fix?

If you did something like:

```ts

// Initialize config
const configuration = {
  param1: BigNumber.from(100),
  param2: BigNumber.from(1000),
};

....
// Log the config
logger.debug({ at: "MyFunction", message: "starting with config", configuration });

...

// Use the config
// This line will error because configuration.param1 is now a string!
if (configuration.param1.gt(currentGasPrice)) {

....
```

File: packages/financial-templates-lib/src/logger/Formatters.ts [hunk 1]
```diff
@@ -14,12 +14,19 @@ export function errorStackTracerFormatter(logEntry: LogEntry) {
...
+  type SymbolRecord = Record<string | symbol, any>;
   try {
-    iterativelyReplaceBigNumbers(logEntry);
+    // Out is the original object if and only if one or more BigNumbers were replaced.
+    const out = iterativelyReplaceBigNumbers(logEntry);
+
+    // Because winston depends on some non-enumerable symbol properties, we explicitly copy those over, as they are not
+    // handled in iterativelyReplaceBigNumbers. This only needs to happen if logEntry is being replaced.
+    if (out !== logEntry)
+      Object.getOwnPropertySymbols(logEntry).map((symbol) => (out[symbol] = (logEntry as SymbolRecord)[symbol]));
```
Comment (mrice32) [inline packages/financial-templates-lib/src/logger/Formatters.ts:25]: I wanted to call this out -- there are some non-enumerable, `Symbol` props that winston depends on (like `Symbol("level")`), so those must be copied over if the object is swapped out.

File: packages/financial-templates-lib/src/logger/Formatters.ts [hunk 2]
```diff
@@ -40,10 +47,20 @@ export function botIdentifyFormatter(botIdentifier: string) {
...
+  // This does a DFS, recursively calling this function to find the desired value for each key.
+  // It doesn't modify the original object. Instead, it creates an array of keys and updated values.
+  const replacements = Object.entries(obj).map(([key, value]): [string, any] => {
+    if (BigNumber.isBigNumber(value) || web3.utils.isBN(value)) return [key, value.toString()];
+    else if (typeof value === "object" && value !== null) return [key, iterativelyReplaceBigNumbers(value)];
+    else return [key, value];
   });
+
+  // This will catch any values that were changed by value _or_ by reference.
+  // If not changes were detected, no copy is needed and it is fine to discard the copy and return the original object.
```
Comment (pxrl) [inline packages/financial-templates-lib/src/logger/Formatters.ts:61]: Very minor: s/not/no ?

**Review by pxrl — APPROVED** [review]
Nice - this looks really good!

**Pull request merged**